### PR TITLE
ManageabilityPkg: Add TransportLib instance for IPMI Serial

### PR DESCRIPTION
**Pull request by gtossk2**
This adds ManageabilityTransportSerialLib instance following the manageability transport framework to support IPMI Serial protocol.

### CODE REVIEW

**Comment by changab** [comment]
Hi @gtossk2, thanks for this contribution.  Lets split this commit into three first. It would give us a clear picture of introducing the serial port interface to edk2 manageability package. 
1. Introduce the new definitions for the serial transport interface. That includes the newly added definitions in the C header file and ManageabilityPkg.dec such as the PCD.
2. We can introduce the new manageability help library for the serial protocol, and add it in the DSC file.
3. The last commit would be adopting the serial protocol in IPMI driver.
Could you please help to update this PR? Thanks!

**Comment by gtossk2** [comment]
Hi @changab Thanks for the suggestion. I have split the commits into three. Please kindly let me know if it is what you expected. Thanks!

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/ManageabilityTransportSerial.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+///
+/// IPMI Serial Configure
+///
+#define IPMI_SERIAL_RS_ADDRESS               FixedPcdGet8  (PcdIpmiSerialResponseAddress)
+#define IPMI_SERIAL_RQ_ADDRESS               FixedPcdGet8  (PcdIpmiSerialRequestAddress)
+#define IPMI_SERIAL_RS_LUN_NUMBER            FixedPcdGet8  (PcdIpmiSerialResponseLunNumber)
+#define IPMI_SERIAL_RQ_LUN_NUMBER            FixedPcdGet8  (PcdIpmiSerialRequestLunNumber)
+#define IPMI_SERIAL_RETRY_COUNT              FixedPcdGet8  (PcdIpmiSerialRetryCount)
+#define IPMI_SERIAL_REQUEST_RETRY_INTERVAL   FixedPcdGet32 (PcdIpmiSerialRequestRetryInterval)
+#define IPMI_SERIAL_RESPONSE_RETRY_INTERVAL  FixedPcdGet32 (PcdIpmiSerialResponseRetryInterval)
```
[the remark this answers is not in the document]
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/ManageabilityTransportSerial.h]: Done. Move the definitions to MdePkg/Include/IndustryStandard. PR: https://github.com/tianocore/edk2/pull/6454

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+static const struct {
+  UINT8    Character;
+  UINT8    Escape;
+} Characters[] = {
+  { BM_START,     0xB0 }, /* start */
+  { BM_STOP,      0xB5 }, /* stop */
+  { BM_HANDSHAKE, 0xB6 }, /* packet handshake */
+  { BM_ESCAPE,    0xBA }, /* data escape */
+  { 0x1B,         0x3B }  /* escape */
+};
```
[the remark this answers is not in the document]
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c]: Acked.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+
+/**
+  Return escaped character for the given one
+
+  @param[in]      Character           Character data.
+
+  @retval         Character           Return escaped character if it is.
+                                      Otherwise, return the given one.
+**/
+static inline UINT8
```
[the remark this answers is not in the document]
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c]: Acked.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+
+/**
+  Return unescaped character for the given one
+
+  @param[in]      Character           Character data.
+
+  @retval         Character           Return unescaped character if it is.
+                                      Otherwise, return the given one.
+**/
+static inline UINT8
```
[the remark this answers is not in the document]
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c]: Acked.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c [hunk 5]
```diff
@@ -0,0 +1,619 @@
...
+  IN OUT    UINT32  *ResponseDataSize
+  )
+{
+  UINT8         *ResponsePtr;
+  UINT8         Character;
+  UINT8         UnescapedCharacter;
+  UINT32        Count;
+  UINT32        Index;
+  STATIC UINT8  CtxState  = MSG_IDLE;
+  STATIC UINT8  CtxEscape = 0;
```
[the remark this answers is not in the document]
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c:108]: Need to keep the value while returning the func. Initialization at the new line will lead to wrong value.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+        if (Character == BM_START) // START
+        {
+            CtxState = MSG_IN_PROGRESS;
+            CtxEscape = 0;
+        }
+        else if (CtxState != MSG_IN_PROGRESS)
+        {
+            continue;
+        }
+        else if (CtxEscape)
```
[the remark this answers is not in the document]
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c]: Acked.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c [hunk 7]
```diff
@@ -0,0 +1,619 @@
...
+      )
+  {
+    DEBUG ((DEBUG_ERROR, "%a: Mismatched values of RequestData or RequestDataSize.\n", __func__));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  // Validation on TransmitHeader and TransmitHeaderSize.
+  if (((TransmitHeader == NULL) && (TransmitHeaderSize != 0)) ||
+      ((TransmitHeader != NULL) && (TransmitHeaderSize == 0))
+      )
```
[the remark this answers is not in the document]
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c:228]: Acked.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c [hunk 8]
```diff
@@ -0,0 +1,619 @@
...
+    if (IpmiSerialGetEscapedCharacter (Buffer[Index]) != Buffer[Index]) {
+      EscapedCharacterCount++;
+    }
+  }
+
+  // Allocate memory buffer include escaped characters
+  RequestLength = BufferLength + EscapedCharacterCount + 2; // start + stop byte
+  Request       = AllocateZeroPool (RequestLength);
+  if (Request == NULL) {
+    DEBUG ((DEBUG_ERROR, "Out Of Resource \n"));
```
[the remark this answers is not in the document]
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c:285]: Free `Request` before returning the func.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c [hunk 9]
```diff
@@ -0,0 +1,619 @@
...
+  //
+  RetryCount = 0;
+  while (TRUE) {
+    Status = SerialPortWrite (Request, RequestLength);
+    if (Status == RequestLength) {
+      break;
+    }
+
+    if (++RetryCount > IPMI_SERIAL_RETRY_COUNT) {
+      DEBUG ((DEBUG_ERROR, "%a: Write Request error %r\n", __FUNCTION__, Status));
```
[the remark this answers is not in the document]
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c]: Acked.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c [hunk 10]
```diff
@@ -0,0 +1,619 @@
...
+  EFI_STATUS  Status;
+
+  if ((DataByte == NULL) || (*Length == 0)) {
+    DEBUG ((DEBUG_ERROR, "%a: Either DataByte is NULL or Length is 0.\n", __func__));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Status = SerialReadResponse (DataByte, Length);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Read Response error %r\n", __FUNCTION__, Status));
```
[the remark this answers is not in the document]
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c]: Acked.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c [hunk 11]
```diff
@@ -0,0 +1,619 @@
...
+
+  if ((ResponseDataSize == NULL) ||
+      (BufferSize > *ResponseDataSize) ||
+      (ResponseData == NULL))
+  {
+    DEBUG ((DEBUG_ERROR, "Response Data insufficient. \n"));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+    ResponsePtr = ResponseData;
```
[the remark this answers is not in the document]
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c]: Acked.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Dxe/ManageabilityTransportSerial.c [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+
+#include "ManageabilityTransportSerial.h"
+
+MANAGEABILITY_TRANSPORT_SERIAL  *mSingleSessionToken = NULL;
+
+EFI_GUID  *SupportedManageabilityProtocol[] = {
+  &gManageabilityProtocolIpmiGuid,
+};
+
+UINT8  NumberOfSupportedProtocol = (sizeof (SupportedManageabilityProtocol)/sizeof (EFI_GUID *));
```
Comment (abdattar) [inline Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Dxe/ManageabilityTransportSerial.c:27]: file global variable should start with small letter 'm'.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Dxe/ManageabilityTransportSerial.c [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+{
+  CHAR16      *ManageabilityProtocolName;
+  EFI_STATUS  Status;
+
+  if (TransportToken == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a: Invalid transport token.\n", __func__));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if (HardwareInfo.Serial == NULL) {
```
Comment (abdattar) [inline Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Dxe/ManageabilityTransportSerial.c:64]: HardwareInfo is marked as OPTIONAL and here accessing it without check for 'NULL'.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Dxe/ManageabilityTransportSerial.c [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+  // Get protocol specification name.
+  ManageabilityProtocolName = HelperManageabilitySpecName (TransportToken->ManageabilityProtocolSpecification);
+
+  DEBUG ((DEBUG_MANAGEABILITY_INFO, "%a: Serial transport hardware for %s is:\n", __func__, ManageabilityProtocolName));
+  DEBUG ((DEBUG_MANAGEABILITY_INFO, "IPMI requester address       : 0x%08x\n", mSerialHardwareInfo.IpmiRequesterAddress));
+  DEBUG ((DEBUG_MANAGEABILITY_INFO, "IPMI responder address       : 0x%08x\n", mSerialHardwareInfo.IpmiResponderAddress));
+  DEBUG ((DEBUG_MANAGEABILITY_INFO, "IPMI requester LUN           : 0x%08x\n", mSerialHardwareInfo.IpmiRequesterLUN));
+  DEBUG ((DEBUG_MANAGEABILITY_INFO, "IPMI responder LUN           : 0x%08x\n", mSerialHardwareInfo.IpmiResponderLUN));
+
+  Status = SerialPortInitialize ();
```
Comment (abdattar) [inline Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Dxe/ManageabilityTransportSerial.c:86]: you can move this line to the begining of the function, if it fails no need to the initialization of mSerialHardwareInfo.

**Review by abdattar — CHANGES_REQUESTED** [review]

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c [hunk 15]
```diff
@@ -0,0 +1,599 @@
...
+#include <Library/ManageabilityTransportMctpLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/TimerLib.h>
+
+#include "ManageabilityTransportSerial.h"
+
+extern MANAGEABILITY_TRANSPORT_SERIAL_HARDWARE_INFO  mSerialHardwareInfo;
+extern MANAGEABILITY_TRANSPORT_SERIAL                *mSingleSessionToken;
+
+UINT8  mNumberOfEscapeCharacters = (sizeof (Characters) / sizeof (Characters[0]));
```
[the remark this answers is not in the document]
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c]: Acked.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Dxe/ManageabilityTransportSerial.c (hunk 12 above)
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Dxe/ManageabilityTransportSerial.c]: Acked.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Dxe/ManageabilityTransportSerial.c (hunk 13 above)
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Dxe/ManageabilityTransportSerial.c:70]: `HardwareInfo` is unios and check `Serial` structure pointer for `NULL`

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/ManageabilityTransportSerial.h [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+#define IPMI_SERIAL_REQUEST_RETRY_INTERVAL   FixedPcdGet32 (PcdIpmiSerialRequestRetryInterval)
+#define IPMI_SERIAL_RESPONSE_RETRY_INTERVAL  FixedPcdGet32 (PcdIpmiSerialRequestRetryInterval)
+
+///
+/// Table of special characters
+///
+static const struct {
+  UINT8    Character;
+  UINT8    Escape;
+} Characters[] = {
```
[the remark this answers is not in the document]
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/ManageabilityTransportSerial.h]: Acked.

**Comment by gtossk2** [comment]
@abdattar Many thanks for the review. I have updated per your comments.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c [hunk 17]
```diff
@@ -0,0 +1,607 @@
...
+    DEBUG ((DEBUG_ERROR, "Response Data insufficient. \n"));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  ResponsePtr = ResponseData;
+  Count       = 0;
+  for (Index = 0; Index < BufferSize; Index++) {
+    Character = Buffer[Index];
+    if (Character == BASIC_MODE_START) {
+      // START
```
Comment (abdattar) [inline Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c:136]: A check is required before assigning to CtxState.
it should be idle state.

**Review by abdattar — CHANGES_REQUESTED** [review]

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c (hunk 17 above)
Comment (gtossk2) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c:138]: Acked.

File: Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c [hunk 18]
```diff
@@ -0,0 +1,609 @@
...
+        CtxState = MSG_IDLE;
+        continue;
+      }
+
+      *ResponsePtr++ = UnescapedCharacter;
+      Count++;
+      CtxEscape = 0;
+    } else if (Character == BASIC_MODE_ESCAPE) {
+      CtxEscape = 1;
+      continue;
```
Comment (abdattar) [inline Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c:157]: Will this cause data loss?
With "continue", the buffer will move ahead, correct?
Do you need to decrement the index?

Comment (abdattar) [reply Features/ManageabilityPkg/Library/ManageabilityTransportSerialLib/Common/SerialCommon.c:157]: Never mind, i got it the next byte is the actual byte to be replaced.

**Comment by abdattar** [comment]
code changes looks good to me.
Please address the single comment mentioned in this review.

**Review by abdattar — APPROVED** [review]

**Review by changab — APPROVED** [review]

**Pull request merged**